File: kafka-sample/python/kafka-oauth-demo.py
#!/usr/bin/env python3
# coding: utf-8

#
# Copyright 2021 Mercedes-Benz Connectivity Services GmbH
#
# SPDX-License-Identifier: MIT

import argparse
import logging

from confluent_kafka import Consumer
from confluent_kafka.cimpl import KafkaException

client_name = 'YOUR_CLIENT_NAME'             # use the client name you have received
client_id = 'YOUR_CLIENT_ID'                 # use the client you have received
client_secret = 'YOUR_CLIENT_SECRET'         # use the secret you have received
scope = 'SCOPE'                              # use the correct scope for your region
topic = f'vehiclesignals.{client_name}'      # use topic for the client you have received
group = f'{client_name}.GROUP_ID_POSTFIX'    # you can change the postfix of your consumer group

bootstrap_url = 'BOOTSTRAP_URL'              # use the correct broker url for your region
oauth_token_api_url = 'OAUTH_TOKEN_API_URL'  # use the correct token API url for your region

security_protocol = 'SASL_SSL'
sasl_mechanism = 'OAUTHBEARER'


def consume_kafka():
    consumer = configure_consumer()
    app_logger.info("Consumer created")
    consumer.subscribe([topic])
    app_logger.info('Subscribed to topic %s', topic)

    app_logger.info("Starting poll loop")
    try:
        while True:
            msg = consumer.poll(timeout=1.0)
            if msg is None:
                app_logger.debug("No new messages")
                continue
            if msg.error():
                raise KafkaException(msg.error())
            else:
                # Process the received message
                print(msg.value().decode('utf-8'))
                # commit errors can be handled with
                # on_commit: Callable[[KafkaError, list[TopicPartition]], None] = commit_cb
                # config['on_commit']=on_commit
                consumer.commit(asynchronous=True)
    except KeyboardInterrupt:
        pass
    finally:
        app_logger.info("shutting down consumer.")
        consumer.close()


def configure_consumer():
    conf = {
        'bootstrap.servers': bootstrap_url,
        'security.protocol': security_protocol,
        'sasl.mechanism': sasl_mechanism,
        'sasl.oauthbearer.method': 'OIDC',
        'sasl.oauthbearer.client.id': client_id,
        'sasl.oauthbearer.client.secret': client_secret,
        'sasl.oauthbearer.scope': scope,
        'sasl.oauthbearer.token.endpoint.url': oauth_token_api_url,
        'group.id': group,
        'auto.offset.reset': 'earliest',
        'enable.auto.commit': True,
    }
    if args.kafka_debug != ["none"]:
        conf['debug'] = 'consumer,security' if args.kafka_debug is None else ','.join(args.kafka_debug)

    consumer_logger = logging.getLogger('consumer')
    handler = logging.StreamHandler()
    handler.setFormatter(logging.Formatter(LOG_FORMAT))
    consumer_logger.addHandler(handler)

    consumer = Consumer(conf, logger=consumer_logger)
    return consumer


def parse_arguments():
    parser = argparse.ArgumentParser(
        prog="oidc-demo",
        description="The Kafka OAuth demo shows how a customer can authenticate their client via OAuth2 to Kafka \
        and how they can read data from a Kafka topic API and using oidc.",
    )
    parser.add_argument(
        '-log',
        '--loglevel',
        default='info',
        help='set root loglevel, default=info'
    )
    parser.add_argument(
        '-kafka-debug',
        nargs='*',
        help='set verbosity of kafka debug logs, disable with -kafka-debug=none. \
        Only useful together with -log=debug. \
        See CONFIGURATION.md in https://github.com/confluentinc/librdkafka for valid `debug` options, \n\
        default=consumer security'
    )
    return parser.parse_args()


if __name__ == '__main__':
    args = parse_arguments()

    LOG_FORMAT = '[%(name)s]: %(asctime)s - %(levelname)s - %(message)s'
    logging.basicConfig(encoding='utf-8', level=args.loglevel.upper(), format=LOG_FORMAT)
    app_logger = logging.getLogger("kafka-oauth-demo")

    consume_kafka()

File: kafka-sample/c#/KafkaOauthDemo.cs
//
// Copyright 2021 Mercedes-Benz Connectivity Services GmbH
//
// SPDX-License-Identifier: MIT

using Confluent.Kafka;

namespace KafkaConsumer
{
    class Program
    {

        static void Main()
        {

            var clientId = "YOUR_CLIENT_ID";                      // If you are an MBCon customer, please use the client id you have received
            var clientSecret = "YOUR_CLIENT_SECRET";              // If you are an MBCon customer, please use the secret you have received
            var scope = "SCOPE";                                  // use the correct scope for your region
            var topic = "YOUR_DEDICATED_TOPIC";                   // If you are an MBCon customer, please use topic name as 'vehiclesignals.<client name>'
            var consumerGroup = "CONSUMER_GROUP";                 // If you are an MBCon customer, please use the received client name as the prefix. eg: '<client name>.GROUP_ID_POSTFIX_OF_YOUR_CHOICE'

            var bootstrapUrl = "BOOTSTRAP_URL";                   // use the correct broker url for your region
            var oauthTokenApiUrl = "OAUTH_TOKEN_API_URL";         // use the correct token API url for your region

            var securityProtocol = SecurityProtocol.SaslSsl;
            var saslMechanism = SaslMechanism.OAuthBearer;
            var sslEndpointIdentificationAlgorithm = SslEndpointIdentificationAlgorithm.None;
            var autoOffsetReset = AutoOffsetReset.Earliest;

            var config = new ConsumerConfig
            {
                BootstrapServers = bootstrapUrl,
                GroupId = consumerGroup,

                SslEndpointIdentificationAlgorithm = sslEndpointIdentificationAlgorithm,
                SecurityProtocol = securityProtocol,
                SaslMechanism = saslMechanism,

                SaslOauthbearerMethod = SaslOauthbearerMethod.Oidc,
                SaslOauthbearerClientId = clientId,
                SaslOauthbearerClientSecret = clientSecret,
                SaslOauthbearerScope = scope,
                SaslOauthbearerTokenEndpointUrl = oauthTokenApiUrl,

                Debug = "consumer,security",
                AutoOffsetReset = autoOffsetReset,
            };


            using var consumer = new ConsumerBuilder<Ignore, string>(config).Build();
            consumer.Subscribe(topic);

            CancellationTokenSource cts = new();
            Console.CancelKeyPress += (_, e) =>
            {
                e.Cancel = true; // prevent the process from terminating.
                cts.Cancel();
            };

            try
            {
                while (true)
                {
                    try
                    {
                        var consumerResult = consumer.Consume(cts.Token);

                        Console.WriteLine($"Consumed message '{consumerResult.Message.Value}' at: '{consumerResult.TopicPartitionOffset}'.");
                    }
                    catch (ConsumeException e)
                    {
                        Console.WriteLine($"Error occurred: {e.Error.Reason}");

                    }
                }
            }
            catch (OperationCanceledException)
            {
                // Ensure the consumer leaves the group cleanly and final offsets are committed.
                consumer.Close();
            }
        }
    }
}
